### DISCUSSION TITLE
Heap size grows when publishing at high frequency

### DISCUSSION THREAD

**Discussion by wangxin5355** [rabbitmq/rabbitmq-dotnet-client#1137] [General]
![image](https://user-images.githubusercontent.com/7795630/151485666-12163e2a-ddf4-46a9-9127-86e156cfc98d.png)
![image](https://user-images.githubusercontent.com/7795630/151485842-40e94ef8-3308-4aa0-9a98-4409714bbc6d.png)
![image](https://user-images.githubusercontent.com/7795630/151485890-256b55aa-8e19-481a-92ee-7cc3e30bca07.png)

**Comment by stebet** [comment]
It's likely that your connection is saturated or your RabbitMQ server is simply not keeping up with publishing.  What happens if you turn on PublisherConfirms and wait for the confirms in each thread instead of doing the `Thread.Sleep(1)`?

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment]
We don't have any version information here.

Publishing in a tight loop is expected to result in (.NET) heap growth up to a point, and RabbitMQ node memory use will also go up up to a point before throttling kicks in. Neither will be perfectly flat in any such scenario (or any scenario at all in a memory managed runtime).

Most of the heap is consumed by a `SingleProducerSIngleConsumerQueue` according to the screenshot. This is not a part of this client as far as GitHub search suggests. So either this is a .NET primitive or something that your own code does not release during the test.
